src/java.base/share/classes/java/io/ProxyingConsole.java

```diff
@@ -0,0 +1,198 @@
+    private final JdkConsole delegate;
+    private final Object readLock;
+    private final Object writeLock;
+    private final Reader reader;
+    private final PrintWriter printWriter;
+
+    ProxyingConsole(JdkConsole delegate) {
+        this.delegate = delegate;
+        readLock = new Object();
+        writeLock = new Object();
+        reader = new WrappingReader(delegate.reader(), readLock);
+        printWriter = new WrappingWriter(delegate.writer(), writeLock);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public PrintWriter writer() {
+        return printWriter;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Reader reader() {
+        return reader;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Console format(String fmt, Object ... args) {
+        synchronized (writeLock) {
+            delegate.format(fmt, args);
+        }
+        return this;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Console printf(String format, Object ... args) {
+        synchronized (writeLock) {
+            delegate.printf(format, args);
+        }
+        return this;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String readLine(String fmt, Object ... args) {
+        synchronized (writeLock) {
+            synchronized (readLock) {
+                return delegate.readLine(fmt, args);
+            }
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String readLine() {
+        synchronized (readLock) {
+            return delegate.readLine();
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public char[] readPassword(String fmt, Object ... args) {
+        synchronized (writeLock) {
+            synchronized (readLock) {
+                return delegate.readPassword(fmt, args);
+            }
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public char[] readPassword() {
+        synchronized (readLock) {
+            return delegate.readPassword();
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void flush() {
+        delegate.flush();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Charset charset() {
+        return delegate.charset();
+    }
+
+    private static class WrappingReader extends Reader {
+        private final Reader r;
+        private final Object lock;
+
+        WrappingReader(Reader r, Object lock) {
+            super(lock);
+            this.r = r;
+            this.lock = lock;
+        }
+
+        @Override
+        public int read(char[] cbuf, int off, int len) throws IOException {
+            synchronized (lock) {
+                return r.read(cbuf, off, len);
+            }
+        }
+
+        @Override
+        public void close() throws IOException {
+            r.close();
```

Comment:
Console specifies that the invoking close on the Reader and Writer does not close the underlying stream. So I think this close (and WrappingWriter::close) need to be a no-op too.

Reply:
Right. Made them as no-op.